stackslib/src/net/inv/nakamoto.rs

```diff
@@ -0,0 +1,215 @@
+    ///
+    /// The resulting bitvector will be truncated if `reward_cycle` is the current reward cycle.
+    pub fn make_tenure_bitvector(
+        &mut self,
+        tip: &BlockSnapshot,
+        sortdb: &SortitionDB,
+        chainstate: &StacksChainState,
+        reward_cycle: u64,
+    ) -> Result<Vec<bool>, NetError> {
+        let ih = sortdb.index_handle(&tip.sortition_id);
+        let reward_cycle_end_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, reward_cycle + 1)
+            - 2;
+        let reward_cycle_end_tip = if tip.block_height <= reward_cycle_end_height {
+            tip.clone()
+        } else {
+            ih.get_block_snapshot_by_height(reward_cycle_end_height)?
+                .ok_or(NetError::NotFoundError)?
+        };
+
+        let mut tenure_status = vec![];
+        let mut cur_height = reward_cycle_end_tip.block_height;
+        let mut cur_consensus_hash = reward_cycle_end_tip.consensus_hash;
+
+        let mut cur_tenure_opt = self.get_processed_tenure(chainstate, &cur_consensus_hash)?;
+
+        loop {
+            let cur_reward_cycle = sortdb
+                .pox_constants
+                .block_height_to_reward_cycle(sortdb.first_block_height, cur_height)
+                .ok_or(NetError::ChainstateError(
+                    "block height comes before system start".into(),
+                ))?;
+            if cur_reward_cycle < reward_cycle {
+                // done scanning this reward cycle
+                break;
+            }
+            let cur_sortition_info = if let Some(info) = self.sortitions.get(&cur_consensus_hash) {
+                info
+            } else {
+                let loaded_info = InvSortitionInfo::load(sortdb, &cur_consensus_hash)?;
+                self.sortitions
+                    .insert(cur_consensus_hash.clone(), loaded_info);
+                self.sortitions
+                    .get(&cur_consensus_hash)
+                    .expect("infallbile: just inserted this data".into())
+            };
+            let parent_sortition_consensus_hash = cur_sortition_info.parent_consensus_hash.clone();
+
+            test_debug!("Get sortition and tenure info for height {}. cur_consensus_hash = {}, cur_tenure_info = {:?}, cur_sortition_info = {:?}", cur_height, &cur_consensus_hash, &cur_tenure_opt, cur_sortition_info);
+
+            if let Some(cur_tenure_info) = cur_tenure_opt.as_ref() {
+                // a tenure was active when this sortition happened...
+                if cur_tenure_info.tenure_id_consensus_hash == cur_consensus_hash {
+                    // ...and this tenure started in this sortition
+                    tenure_status.push(true);
+                    cur_tenure_opt = self.get_processed_tenure(
+                        chainstate,
+                        &cur_tenure_info.parent_tenure_id_consensus_hash,
+                    )?;
```

Comment:
is `cur_tenure_info.parent_tenure_id_consensus_hash` _always_ equal to `parent_sortition_consensus_hash`? I don't think it is... for example, if the prior bitcoin block didn't produce a sortition winner. 

In which case, I think this loop's variable maintenance would be broken: `cur_height` would decrement by 1, `cur_consensus_hash` would point at the parent bitcoin block, and `cur_tenure_opt` would be a tenure starting at a different bitcoin block.

My suggestion would be to move `cur_tenure_opt`'s binding into the loop's scope, and then the start of the loop initializes  cur_tenure_opt:

```rust
loop {
    ...
    if cur_reward_cycle < reward_cycle {
        break;
    }
    let cur_tenure_opt = self.get_processed_tenure(chainstate, &cur_consensus_hash);
```

I think this makes the loop easier to understand, and eliminates code. Because the tenure_status could just be:

```rust
// we've processed the tenure if we have information for it, and the tenure began in this sortition.
let tenure_status = matches!(
    cur_tenure_opt.as_ref(), 
    Some(cur_tenure_info) if cur_tenure_info.tenure_id_consensus_hash == cur_consensus_hash
);
tenure_statuses.push(tenure_status);
```

Or something along those lines if you don't prefer to use `matches!`

Reply:
> is cur_tenure_info.parent_tenure_id_consensus_hash always equal to parent_sortition_consensus_hash? I don't think it is... for example, if the prior bitcoin block didn't produce a sortition winner.

Yes, correct. That's accounted for in the `else`-branch of this `if/else` block.  This is also tested via the `test_nakamoto_inv_2_tenures_3_sortitions` unit tests.  You can see that we get a `false` for one of the sortitions, since it didn't have a corresponding tenure.

Reply:
The problem is this part of the code:

```rust
            if let Some(cur_tenure_info) = cur_tenure_opt.as_ref() {
                // a tenure was active when this sortition happened...
                if cur_tenure_info.tenure_id_consensus_hash == cur_consensus_hash {
                    // ...and this tenure started in this sortition
                    tenure_status.push(true);
                    cur_tenure_opt = self.get_processed_tenure(
                        chainstate,
                        &cur_tenure_info.parent_tenure_id_consensus_hash,
                    )?;
  ...
```

Combined with this part of the code:

```rust
            cur_consensus_hash = parent_sortition_consensus_hash;
            if cur_height == 0 {
                break;
            }
            cur_height = cur_height.saturating_sub(1);
```

So, for example, if the bitcoin blocks 1 & 3 contained successful sortitions but 2 didn't, after evaluating block 3, the loop would set:

`cur_consensus_hash = bitcoin block 2` (because `parent_consensus_hash` of 3 is 2).
`cur_tenure_opt = Some(tenure of bitcoin block 1)` (because the `parent_tenure_id_consensus_hash` of 3 would be 1)
`cur_height = 2`

So, the three loop variables would be out of sync. Maybe this doesn't create an issue in tests because `cur_tenure_info.consensus_hash != consensus_hash`, so it pushes `tenure_status = false` anyways, but it should be observable at some point, because the loop variables do appear to get unsynced.

Reply:
> So, the three loop variables would be out of sync.

No, that's deliberate.  `cur_height` tracks the position in the inventory vector, which gets set to `true` or `false` based on whether or not `cur_tenure_opt` matches the consensus hash of the sortition corresponding to `cur_height`.

Basically, this loop checks each sortition at every height in the reward cycle, and determines from there whether or not the sortition was the start of a tenure. If so, then `true` gets pushed to the inventory vector. Otherwise, `false` will be pushed.

I've significantly increased test coverage to synthesize multi-reward-cycle chain histories from inventory vectors, and then verified that the inventory generation logic reproduces the inventory vectors from the synthesized chain history.